### PR TITLE
[FSDP] Add no_sync() context manager

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72578 [ZeRO] Add ctor support for multiple param groups
* **#72446 [FSDP] Add no_sync() context manager**

**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun4 python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    @parametrize(
+        "num_iters_to_acc",
+        [2, 3, 4],
+    )
+    def test_no_sync(self, num_iters_to_acc):
+        assert num_iters_to_acc >= 2, \
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(batch_dim=1, num_iters_to_acc=num_iters_to_acc)
+
+    def _test_communication(self):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:115]: We can test the communication costs with benchmarks, is this test necesssary / does it add value in terms of testing the API?

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestGradAcc(FSDPTest):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:32]: Can we test with CPU offload, backward prefetch, etc as well just to make sure things are working as expected?

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        fsdp_model.zero_grad()
+        losses = []
+        with fsdp_model.no_sync():
+            for batch in batches[:-1]:  # accumulate for all but the last batch
+                output = fsdp_model(*batch)
+                loss = fsdp_model.module.get_loss(batch, output)
+                loss.backward()
+                losses.append(loss)
+        output = fsdp_model(*batches[-1])
+        loss = fsdp_model.module.get_loss(batches[-1], output)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:90]: Should we run an optimizer.step() after no_sync() is exited just to make sure things work fine?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
...
+        old_flags = []
+        for m in self.modules():
+            if isinstance(m, FullyShardedDataParallel):
+                old_flags.append((m, m._require_backward_grad_sync))
+                m._require_backward_grad_sync = False
+        try:
+            yield
+        finally:
+            for m, old_flag in old_flags:
+                assert not m._require_backward_grad_sync
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1130]: Add an error msg?

**Review by rohan-varma — COMMENTED** [review]
Oops didn't see this was a draft and I already wrote some comments, so just sending them your way for consideration!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -810,7 +813,12 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 "FSDP only works with gradients that don't require gradients"
             )
 
-        self._free_full_params([param])
+        if self._require_backward_grad_sync:
+            # We do not free the full parameters in a `no_sync()` context since
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: I realized that this was legacy for the Fairscale implementation and was removed in more recent commits. I have removed this conditional.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
...
+            networking overhead.
+        """
+        self._lazy_init()
+        assert self._is_root, \
+            "`no_sync()` on inner FSDP instances is not supported"
+        self._assert_state(TrainingState_.IDLE)
+        old_flags = []
+        for m in self.modules():
+            if isinstance(m, FullyShardedDataParallel):
+                old_flags.append((m, m._require_backward_grad_sync))
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1120]: This is minor, but in my understanding, Python generator ordering is not guaranteed. In practice, `torch.nn.Module.modules()` probably does always return the same ordering, but this might be relying on non-specified behavior.

Is there any downside to maintaining an extra reference to each module? The modules should not be being deleted anyway (so these references should not be the critical ones preventing garbage collection)?

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py]: This test is pretty interesting to me. When you wrote this comment, I had not yet pushed what I had, so it probably did not make much sense.

For example, through this test, I was able to see that PyTorch `FSDP` makes an additional all-gather call in the forward pass compared to Fairscale `FSDP`. Also, only thinking about the specification of `no_sync()`, I think testing that it indeed saves on communication primitives makes sense.

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 2 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:32]: Testing with those orthogonal features makes sense to me! (though I may need some time to familiarize myself)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
...
+        A context manager to disable gradient synchronizations across FSDP
+        instances. Within this context, gradients will be accumulated on module
+        variables, which will later be synchronized in the first
+        forward-backward pass after exiting the context.
+
+        .. note:: This likely results in higher memory usage because FSDP will
+            accumulate the full model gradients (instead of gradient shards)
+            until the eventual sync.
+
+        .. note:: Gradient accumulation can be done without this context,
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1109]: Good catch.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
                 traceback.print_stack()
             raise ValueError(msg)
 
+    @contextmanager
+    def no_sync(self) -> Generator:
+        """
+        A context manager to disable gradient synchronizations across FSDP
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1100]: Done.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 4 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1126]: Done.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 9]
```diff
@@ -9,6 +10,7 @@
     Callable,
     cast,
     Dict,
+    Generator,
     List,
     Optional,
     Generator,
```
Comment (mrshenli) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: lint error is real

**Review by mrshenli — COMMENTED** [review]
A meta question:

`no_sync` is certainly useful feature for users, but IIUC it is not very friendly to FSDP, as we will need to either run H2D/D2H and allgather to collect parameters for every micro-batch (costly) or keep params + grads in CUDA memory (won't work for large models). 

I wonder if there are plans to add real micro-batching for FSDP? I.e., FSDP still takes one mini-batch, and then each FSDP unit chunk the input into multiple micro-batches and process all of them one-by-one before moving on to the next FSDP unit. More specifically, 

current `no_sync` does: FSDP_unit1(micro_batch1) -> FSDP_unit2(micro_batch1) -> FSDP_unit1(micro_batch2) -> FSDP_unit2(micro_batch2)

real micro-batching does : FSDP_unit1(micro_batch1) -> FSDP_unit1(micro_batch2) -> FSDP_unit2(micro_batch1) -> FSDP_unit2(micro_batch2)

The benefit of real micro-batching is that we only need to run one D2H/H2D + allgather across all micro-batches for each FSDP unit's parameters, rather than running D2H/H2D + allgather once for every micro-batch.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: Good catch. My VSCode Python extension broke starting yesterday, and I was finally able to fix it by reverting to an older version. :sweat_smile:

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        torch.testing.assert_allclose(ref_loss, acc_loss)
+        assert len(ref_grads) == len(acc_grads)
+        for ref_grad, acc_grad in zip(ref_grads, acc_grads):
+            assert ref_grad.device == acc_grad.device
+            assert ref_grad.size() == acc_grad.size()
+            assert ref_grad.dtype == acc_grad.dtype
+            torch.testing.assert_allclose(ref_grad, acc_grad)
+
+        # Restore the default dtype
+        torch.set_default_dtype(old_default_dtype)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:115]: Might want to do a try: finally block to ensure this gets reset appropriately even if the test somehow fails

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                group, cuda_first=False,
+            )
+        batch = fsdp_model.module.get_input(device)
+
+        num_fsdp = 0
+        for m in fsdp_model.modules():  # includes `self`
+            if isinstance(m, FSDP) and len(m.params) > 0:
+                num_fsdp += 1
+
+        # Not all PyTorch versions have the `_base` methods available
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:168]: I'm not sure if we need this as we don't test against all PyTorch versions, only PT master which should have both of these. In PT FSDP codebase I believe we unconditionally use them too

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 10 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:131]: Great point!

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 11 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py]: Yup, fixed.

File: test/distributed/fsdp/test_fsdp_grad_acc.py (hunk 3 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_grad_acc.py:112]: Added.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

File: test/distributed/fsdp/test_fsdp_grad_acc.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        self,
+        num_iters_to_acc: int,
+        cpu_offload: CPUOffload,
+        backward_prefetch: Optional[BackwardPrefetch],
+    ):
+        """Tests the ``no_sync()`` context manager."""
+        assert num_iters_to_acc >= 2, \
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(
+            batch_dim=1,
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_grad_acc.py:157]: The only purpose of `test_no_sync()` is to manually pass in `batch_dim=1` to `_test_no_sync()`. If this seems unnecessary, I can coalesce the two methods.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Comment by awgu** [comment]
I changed the test filename from `test_fsdp_grad_acc.py` to `test_fsdp_no_sync.py` since the only way to do gradient accumulation at this point is to use `no_sync()`. Fairscale seems to have another way without `no_sync()` (see https://github.com/pytorch/pytorch/issues/72185). We can rename the file and test class if/when we upstream that.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 13]
```diff
@@ -1175,6 +1181,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
                 traceback.print_stack()
             raise ValueError(msg)
 
+    @contextmanager
+    def no_sync(self) -> Generator:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1185]: Can we also file an issue to refactor/deduplicate the no_sync in DDP and FSDP? They look pretty similar and we can probably provide some sort of config / lambda to take care of DDP and FSDP implementation specific details.

Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1185]: https://github.com/pytorch/pytorch/issues/72589

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Review by zhaojuanmao — COMMENTED** [review]
Thanks for adding this! the code changes look good to me, will check the test codes later on

File: test/distributed/fsdp/test_fsdp_comm.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
+# Owner(s): ["oncall: distributed"]
```
Comment (zhaojuanmao) [inline test/distributed/fsdp/test_fsdp_comm.py:1]: nit: could we rename this test file name to be "test_fsdp_no_sync_comm" or merge with "test_fsdp_no_sync"?  as both of them are testing no_sync features

**Review by zhaojuanmao — COMMENTED** [review]
solid and clean tests! look good from my side, just one nit. Will let @rohan-varma accept it, thanks!

File: test/distributed/fsdp/test_fsdp_comm.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    )
+    def test_communication(
+        self,
+        nested_model: bool,
+        use_no_sync: bool,
+        num_sync_iters: int,
+        num_no_sync_iters: int,
+    ):
+        """
+        Tests the ``no_sync()`` context manager by checking the communication
```
Comment (zhaojuanmao) [inline test/distributed/fsdp/test_fsdp_comm.py:57]: nit: change the comment to mention that this test is generally for checking the communication cost

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_comm.py:57]: Great catch. I see why it was confusing. I will update this.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Review by rohan-varma — APPROVED** [review]
Awesome work and great tests!

File: test/distributed/fsdp/test_fsdp_comm.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestCommunication(FSDPTest):
+    """Tests ``FullyShardedDataParallel``'s collective communication usage."""
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_comm.py:33]: Good point. I will reduce the number of tests to 4 by always using `num_no_sync_iters = 3` and `num_sync_iters = 3`. I had both being either `1` or `3` since we originally suspected that there may be different behavior depending on how many iterations were run.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).
